apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerParameterCollection.cs

```diff
@@ -0,0 +1,196 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Linq;
+using Google.Protobuf.Collections;
+using Google.Protobuf.WellKnownTypes;
+
+namespace Google.Cloud.Spanner
+{
+    /// <summary>
+    /// </summary>
+    public sealed class SpannerParameterCollection : DbParameterCollection
+    {
+        /// <inheritdoc />
+        public override int Count => InnerList.Count;
+
+        /// <inheritdoc />
+        public override object SyncRoot => InnerList;
+
+        private List<SpannerParameter> InnerList { get; } = new List<SpannerParameter>();
+
+        /// <summary>
+        /// </summary>
+        /// <param name="parameterName"></param>
+        /// <param name="dbType"></param>
+        public void Add(string parameterName, SpannerDbType dbType)
+        {
+            InnerList.Add(new SpannerParameter(parameterName, dbType));
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="parameterName"></param>
+        /// <param name="value"></param>
+        /// <param name="dbType"></param>
+        public void Add(string parameterName, object value, SpannerDbType dbType)
+        {
+            InnerList.Add(new SpannerParameter(parameterName, dbType) {Value = value});
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="parameterName"></param>
+        /// <param name="dbType"></param>
+        /// <param name="sourceColumn"></param>
+        public void Add(string parameterName, SpannerDbType dbType, string sourceColumn)
+        {
+            InnerList.Add(new SpannerParameter(parameterName, dbType, sourceColumn));
+        }
+
+        /// <inheritdoc />
+        public override int Add(object value)
+        {
+            InnerList.Add((SpannerParameter) value);
+            return InnerList.Count - 1;
+        }
+
+        /// <inheritdoc />
+        public override void AddRange(Array values)
+        {
+            InnerList.AddRange(values.Cast<SpannerParameter>());
+        }
+
+        /// <inheritdoc />
+        public override void Clear()
+        {
+            InnerList.Clear();
+        }
+
+        /// <inheritdoc />
+        public override bool Contains(object value)
+        {
+            return InnerList.Contains(value as SpannerParameter);
+        }
+
+        /// <inheritdoc />
+        public override bool Contains(string value)
+        {
+            return InnerList.Any(x => x.ParameterName.Equals(value));
+        }
+
+        /// <inheritdoc />
+        public override void CopyTo(Array array, int index)
+        {
+            foreach (var item in InnerList)
```

Comment:
Is it okay for CopyTo to modify the array even if it's going to throw an exception due to running off the end? (And isn't there a CopyTo method that would do this better?)